### DISCUSSION TITLE
Rhai shell?

### DISCUSSION THREAD

**Discussion by lasiotus** [rhaiscript/rhai#679] [Ideas]
Hello!

I've been looking for a shell implemented in Rust that does not have libc/Linux/etc. dependencies. I.e. Rust std only. Although there are things like nushell, they all seem to heavily integrate with existing OSes, so they cannot be easily used with a new OS/kernel.

It looks like Rhai + rhai-fs can be used as the base for such a shell, although things like spawning processes, streams/piping and similar appear to be missing. And, of course, basic terminal integration.

Is this something that somebody has been thinking about and/or working on?

Thanks!

**Comment by schungx** [comment]
I suppose the reason why you'd like to use Rhai for your shell is to make it scriptable?

In this case, you'd need to have some way to clearly distinguish between a normal command line vs a Rhai script.

If you can do that cleanly, I can't see much difficulty in implementing one, although you're probably going to need all the standard goodies (such as line editing, history, completion, coloring, prompts and what not) with your shell... so in the end, won't your shell be just as large as the other?

**Reply by lasiotus** [reply]
Yea, I suppose a shell scripting language should have a concept of line-by-line execution, and Rhai is not that kind of scripting language. Still, I feel there should be a good use for Rhai in a "Rust-first" OS...

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.
